[[ LCB ]] Add varargs support to foreign handlers

runrevmark opened the pull request:
This patch adds varargs support to foreign C handlers, allowing
binding to a wider range of C functions.

The syntax for vararg handlers is similar to that of C:
```
  foreign handler myvarargs(in pFirst as Pointer, ...) returns CInt binds to ...
```
Using the token `...` to indicate the start of the variable
parameter list.

To implement vararg functions, a new parameter mode 'variadic'
has been added which causes a handler to be treated as variadic if
such a parameter with a mode is present.

As C requires any int type with rank less than int to be promoted
to int, and float to be promoted to double (when passed to a non-fixed
parameter in a variadic function), a 'promotedtype' and 'promote'
function have been added to foreign type descriptors to describe
the appropriate relation.

runrevmark committed:
[[ LCB ]] Add varargs support to foreign handlers

This patch adds varargs support to foreign C handlers, allowing
binding to a wider range of C functions.

The syntax for vararg handlers is similar to that of C:
```
  foreign handler myvarargs(in pFirst as Pointer, ...) returns CInt binds to ...
```
Using the token `...` to indicate the start of the variable
parameter list.

To implement vararg functions, a new parameter mode 'variadic'
has been added which causes a handler to be treated as variadic if
such a parameter with a mode is present.

As C requires any int type with rank less than int to be promoted
to int, and float to be promoted to double (when passed to a non-fixed
parameter in a variadic function), a 'promotedtype' and 'promote'
function have been added to foreign type descriptors to describe
the appropriate relation.

livecodeali commented on libscript/src/script-execute.cpp:
```diff
@@ -289,6 +320,219 @@ __MCScriptComputeSlotAttributes(const MCResolvedTypeInfo& p_slot_type,
+{
+    // Get the value in the register, this determines the type.
+    MCValueRef t_arg_value = nil;
+    t_arg_value = CheckedFetchRegister(p_arg_reg);
```
This should probably just be `MCValueRef t_arg_value = CheckedFetchRegister(p_arg_reg);`

livecodeali commented on libfoundation/src/foundation-typeinfo.cpp:
```diff
@@ -803,6 +805,14 @@ static bool MCCommonHandlerTypeInfoCreate(bool p_is_foreign, const MCHandlerType
+        
+        if (p_fields[i].mode == kMCHandlerTypeFieldModeVariadic)
+        {
+            p_field_count = i;
```
Perhaps it's worth asserting that `p_field_count == i + 1` and `i != 0` as these things are checked at (lc-)compile-time aren't they? Just to ensure we don't break the assumptions about the position of the variadic param internally anywhere.

livecodeali commented on libscript/src/script-execute.cpp:
```diff
@@ -202,10 +209,32 @@ class MCScriptForeignInvocation
             break;
         
         case kMCScriptForeignHandlerLanguageC:
```
We should probably throw an error if a variadic foreign handler binds to java since it's not yet supported

runrevmark replied:
I fixed this in the MCJavaCheckSignature - it throws a binding error if there is a variadic parameter present.

runrevmark committed:
[[ LCB ]] Ensure an error is thrown if binding varargs to Java

This patch makes it so that JavaCheckSignature will fail if
an attempt is made to bind to a variadic signature.

runrevmark committed:
[[ LCB ]] Ensure variadic parameter position is checked

This patch ensures that the position of the variadic parameter
is valid when creating a handler typeinfo. If it is not, the
creation fails, and an error is thrown.

runrevmark committed:
[[ LCB ]] Tidy up source

This patch tidies up a case of variable definition and
initialization being in separate commands.

runrevmark commented:
@livecodeali Hopefully that's all the current feedback addressed appropriately.

livecodeali commented:
@livecode-vulcan review ok a8b2d34

livecode-vulcan commented:
💙 review by @livecodeali ok a8b2d34de3bdbb854b83940179ed7cebe34a7457

livecode-vulcan commented:
😎 test success a8b2d34de3bdbb854b83940179ed7cebe34a7457

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/458)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/466)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/445)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/388)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/390)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/453)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/458)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/456)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/470)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/456)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/465)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/483)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/452)
